service/forward_service.hh

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (C) 2021-present ScyllaDB
+ */
+
+/*
+ * This file is part of Scylla.
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <seastar/core/sharded.hh>
+#include <seastar/core/distributed.hh>
+#include <seastar/core/future.hh>
+
+#include "query-request.hh"
+#include "database_fwd.hh"
+#include "message/messaging_service_fwd.hh"
+#include "locator/token_metadata.hh"
+
+namespace service {
+
+class forward_service : public seastar::peering_sharded_service<forward_service> {
+private:
+    netw::messaging_service& _messaging;
+    distributed<database>& _db;
+    const locator::shared_token_metadata& _shared_token_metadata;
+
+public:
+    forward_service(netw::messaging_service& ms, distributed<database> &db,
+        const locator::shared_token_metadata& stm)
+        : _messaging(ms)
+        , _db(db)
+        , _shared_token_metadata(stm) {
+    }
+
+    future<> stop();
+    void init_messaging_service();
+    future<> uninit_messaging_service();
+
+    /* Splits given `forward_request` and distributes execution of resulting
+     * subrequests across a cluster
+     */
+    future<query::forward_result> dispatch(query::forward_request req);
+
+private:
+    /* Used to execute a `forward_request` on remote node
+     */
+    future<query::forward_result> execute(query::forward_request req);
+
+    locator::token_metadata_ptr get_token_metadata_ptr() const noexcept;
+};
```

Comment:
Consider splitting the client and server parts into separate services, if it helps with the dependencies.

Reply:
ok